lib/Parse/ParseExpr.cpp

```diff
@@ -2806,14 +2811,22 @@ ExprResult Parser::ParseInteropTypeAnnotation(const Declarator &D, bool IsReturn
 
 ExprResult Parser::ParseBoundsExpressionOrInteropType(const Declarator &D,
                                                       bool IsReturn) {
-  if (StartsBoundsExpression(Tok))
-    return ParseBoundsExpression();
+  ExprResult Result(true);
+  bool isBoundsorItype = false;
 
-  if (StartsInteropTypeAnnotation(Tok))
-    return ParseInteropTypeAnnotation(D, IsReturn);
+  if (StartsBoundsExpression(Tok)) {
+    Result = ParseBoundsExpression();
+    isBoundsorItype = true;
+  }
 
-  Diag(Tok, diag::err_expected_bounds_expr_or_interop_type);
-  return ExprError();
+  if (StartsInteropTypeAnnotation(Tok)) {
+    Result = ParseInteropTypeAnnotation(D, IsReturn);
+    isBoundsorItype = true;
+  }
+
+  SkipInvalidExprAndParseRelativeBoundsClause(Result, isBoundsorItype);
```

Comment:
I think improving error recovery for bounds expressions is outside the scope of this pull request.  We should just be focused on parsing rel_align in this pull request.  Yes, a parsing error in a bounds expression can lead to poor handling of rel_align.   That is OK, we can fix that in a future check-in. 

 I'd like you to change the code as follows:
- Please remove SkipInvalidAndParseRelativeBoundsClause from this pull request.
- Instead, if the next token starts a relative bounds clause, call ParseRelativeBoundsClause.
- Otherwise,
-- if isBoundOrItype is false, emit an error message stating that we expected a bounds expression or interop type here.
-- if isBoundsOrIType is true, emit a different error message stating the token is unexpected; we expected rel_align or rel_align_value.
- if the token starts a relative .